cluster-operator/src/main/java/io/strimzi/operator/cluster/model/nodepools/NodeIdAssignor.java

```diff
@@ -116,32 +133,175 @@ public NodeIdAssignment assignmentForPool(String poolName)   {
+     *
+     * @param idsInUse      Set with currently used IDs
+     * @param nextIdRange   The NodeIdRange which will pick up the next node ID
+     *
+     * @return  Next unused node ID from the range or null if no such ID was found
+     */
+    private Integer nextUnusedIdFromRange(TreeSet<Integer> idsInUse, NodeIdRange nextIdRange)   {
+        Integer nextId = null;
+
+        if (nextIdRange != null)    {
+            while ((nextId = nextIdRange.getNextNodeId()) != null)  {
+                // If the next ID is not in use, we can break the while loop
+                // If it is already used, we continue wth the next ID
+                if (!idsInUse.contains(nextId)) {
+                    break;
+                }
+            }
         }
 
-        idsInUse.add(nextId);
         return nextId;
     }
+
+    /**
+     * Creates the Node ID range based on the strimzi.io/next-node-ids annotation on the Node Pool.
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param pool              NodePool instance
+     *
+     * @return  The NodeIdRange based on the annotation or null when the annotation is not set or is invalid.
+     */
+    private NodeIdRange nextIdRange(Reconciliation reconciliation, KafkaNodePool pool)   {
+        String nextNodeIds = Annotations.stringAnnotation(pool, Annotations.ANNO_STRIMZI_IO_NEXT_NODE_IDS, null);
+
+        try {
+            return nextNodeIds == null ? null : new NodeIdRange(nextNodeIds);
+        } catch (NodeIdRange.InvalidNodeIdRangeException e) {
+            LOGGER.warnCr(reconciliation, "Failed to use " + Annotations.ANNO_STRIMZI_IO_NEXT_NODE_IDS + " " + nextNodeIds + ". New node IDs will be assigned sequentially.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Generates the next node ID which should be removed based on the already used Node IDs and the user preferred range.
+     *
+     * @param desired           TreeSet with already used IDs. This should contain all IDs currently used (including those which
+     *                          will be scaled done) or those already generated. The newly generated ID is automatically added
+     *                          to this set by this method.
+     * @param removeIdRange     The NodeIdRange which will pick up the next node ID
+     *
+     * @return  The next node ID to remove in scale down
+     */
+    private int nextRemoveId(TreeSet<Integer> desired, NodeIdRange removeIdRange) {
+        // We try to first get the Node ID from range first
+        Integer removeId = removeUsedIdFromRange(desired, removeIdRange);
+
+        // We did not get an ID from the ID range. So we get the next ID only
+        if (removeId == null) {
+            removeId = desired.pollLast();
+        }
+
+        return removeId;
+    }
+
+    /**
+     * Finds the next node ID from a range which is not in use. The IDs are found from range specified in the Node Pool
+     * annotation strimzi.io/next-node-ids.
```

Comment:
I think this Javadoc is wrong. It has to refer the id to be removed or? So even the `strimzi.io/remove-node-ids` annotation.